control-plane-pki-operator/certificaterevocationcontroller/certificaterevocationcontroller.go

```diff
@@ -0,0 +1,683 @@
+	}
+}
+
+func configMapForSignerClass(namespace string, signer certificates.SignerClass) (*corev1.ConfigMap, bool) {
+	switch signer {
+	case certificates.CustomerBreakGlassSigner:
+		return manifests.CustomerSystemAdminSignerCA(namespace), true
+	default:
+		return nil, false
+	}
+}
+
+func (c *CertificateRevocationController) syncCertificateRevocationRequest(ctx context.Context, syncContext factory.SyncContext) error {
+	namespace, name, err := cache.SplitMetaNamespaceKey(syncContext.QueueKey())
+	if err != nil {
+		return err
+	}
+
+	action, requeue, err := c.processCertificateRevocationRequest(namespace, name, nil)
+	if err != nil {
+		return err
+	}
+	if requeue {
+		return factory.SyntheticRequeueError
+	}
+	if action != nil {
+		if err := action.validate(); err != nil {
+			panic(err)
+		}
+		if action.event != nil {
+			syncContext.Recorder().Eventf(action.event.reason, action.event.messageFmt, action.event.args...)
+		}
+
+		if action.crr != nil {
+			_, err := c.hypershiftClient.HypershiftV1beta1().CertificateRevocationRequests(*action.crr.Namespace).ApplyStatus(ctx, action.crr, metav1.ApplyOptions{FieldManager: c.fieldManager})
+			return err
+		} else if action.cm != nil {
+			_, err := c.kubeClient.CoreV1().Secrets(*action.secret.Namespace).Apply(ctx, action.secret, metav1.ApplyOptions{FieldManager: c.fieldManager})
+			return err
+		} else if action.secret != nil {
+			_, err := c.kubeClient.CoreV1().ConfigMaps(*action.cm.Namespace).Apply(ctx, action.cm, metav1.ApplyOptions{FieldManager: c.fieldManager})
+			return err
+		}
+	}
+
+	return nil
+}
+
+type actions struct {
+	event  *eventInfo
+	crr    *hypershiftv1beta1applyconfigurations.CertificateRevocationRequestApplyConfiguration
+	secret *corev1applyconfigurations.SecretApplyConfiguration
+	cm     *corev1applyconfigurations.ConfigMapApplyConfiguration
+}
+
+func (a *actions) validate() error {
+	var set int
+	if a.crr != nil {
+		set += 1
+	}
+	if a.cm != nil {
+		set += 1
+	}
+	if a.secret != nil {
+		set += 1
+	}
+	if set != 1 {
+		return errors.New("programmer error: more than one action set")
+	}
+	return nil
+}
+
+type eventInfo struct {
+	reason, messageFmt string
+	args               []interface{}
+}
+
+func event(reason, messageFmt string, args ...interface{}) *eventInfo {
+	return &eventInfo{
+		reason:     reason,
+		messageFmt: messageFmt,
+		args:       args,
+	}
+}
+
+func (c *CertificateRevocationController) processCertificateRevocationRequest(namespace, name string, now func() time.Time) (*actions, bool, error) {
+	if now == nil {
+		now = time.Now
+	}
+
+	crr, err := c.getCRR(namespace, name)
+	if apierrors.IsNotFound(err) {
+		return nil, false, nil // nothing to be done, CSR is gone
+	}
+	if err != nil {
+		return nil, false, err
+	}
+
+	switch crr.Status.Phase {
+	case hypershiftv1beta1.CertificateRevocationRequestPhaseUnknown, "":
+		// validate the signer class - we've got a validation rule on the object, but system:masters can bypass that
```

Comment:
API validation cannot be bypassed by system:masters.  An admission check somewhere?

Reply:
ACK, then I got my wires crossed and this should be impossible.